File: 2015/ruby/day9.rb
def assert_equal(expected, actual, str="")
  raise "Expected #{expected} but got #{actual}. #{str}" unless expected == actual
end
data = File.read('day9.txt')

sample = <<~TXT
London to Dublin = 464
London to Belfast = 518
Dublin to Belfast = 141
TXT

def distance(routes, start, destinations, type = :min)
  raise 'Doh..no destinations' if destinations.length == 0
  return routes[start][destinations[0]] if destinations.length == 1

  distances = destinations.map do |destination|
    routes[start][destination] + distance(routes, destination, destinations - [destination], type)
  end

  type == :min ? distances.min : distances.max
end

def load_routes(data)
  routes = {}
  data.lines.each do |line|
    locs, dist = line.strip.split(' = ')
    l1, l2 = locs.split(' to ')
    routes[l1] ||= {}
    routes[l2] ||= {}
    routes[l1][l2] = dist.to_i
    routes[l2][l1] = dist.to_i
  end
  routes  
end

def shortest(data)
  routes = load_routes(data)
  routes.keys.map do |start|
    distance(routes, start, routes.keys - [start])
  end.min
end

def longest(data)
  routes = load_routes(data)
  routes.keys.map do |start|
    distance(routes, start, routes.keys - [start], :max)
  end.max
end

assert_equal(605, shortest(sample))
puts "Part 1: #{shortest(data)}"

assert_equal(982, longest(sample))
puts "Part 2: #{longest(data)}"

File: 2015/go/day9/day9.go
package day9

import "fmt"

func Shortest(input []string) int {
	return ShortLong(input, "min")
}

func Longest(input []string) int {
	return ShortLong(input, "max")
}

func ShortLong(Input []string, LookupType string) int {
	Routes := Parse(Input)
	Destinations := make([]string, 0)

	for k := range Routes {
		Destinations = append(Destinations, k)
	}
	Distances := make([]int, len(Destinations))
	for Idx, Start := range Destinations {
		Distances[Idx] = Distance(Routes, Start, OtherThan(Destinations, Start), LookupType)
	}
	min, max := MinMax(Distances)
	if LookupType == "min" {
		return min
	} else {
		return max
	}
}

func OtherThan(All []string, Minus string) []string {
	Copy := make([]string, len(All))
	copy(Copy, All)
	for Idx, Val := range Copy {
		if Val == Minus {
			return append(Copy[:Idx], Copy[Idx+1:]...)
		}
	}
	return Copy
}

func Distance(Routes map[string]map[string]int, Start string, Destinations []string, LookupType string) int {
	if len(Destinations) == 1 {
		return Routes[Start][Destinations[0]]
	}
	Distances := make([]int, 0)
	for _, Destination := range Destinations {
		OtherDestinations := OtherThan(Destinations, Destination)
		Distances = append(Distances, Routes[Start][Destination]+Distance(Routes, Destination, OtherDestinations, LookupType))
	}
	min, max := MinMax(Distances)
	if LookupType == "min" {
		return min
	} else {
		return max
	}
}

func MinMax(vals []int) (int, int) {
	min := vals[0]
	max := vals[0]
	for _, v := range vals {
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
	}
	return min, max
}

func Parse(input []string) map[string]map[string]int {
	rv := map[string]map[string]int{}

	for _, line := range input {
		var from, to string
		var dist int
		fmt.Sscanf(line, "%s to %s = %d", &from, &to, &dist)

		_, ok := rv[from]
		if !ok {
			rv[from] = map[string]int{}
		}
		_, ok = rv[to]
		if !ok {
			rv[to] = map[string]int{}
		}

		rv[from][to] = dist
		rv[to][from] = dist
	}
	return rv
}
